### ISSUE TITLE
Enhancement: explicit-function-return-type - Allow omitting return type declaration for inferred void return values

### ISSUE THREAD

**Issue by KristofferHenriksson** [typescript-eslint/typescript-eslint#5469]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/explicit-function-return-type

### Description

I propose that an option be added to allow omitting the return type declaration if the function can be inferred to return void. This seems to be a common pattern with developers only explicitly writing return types when there's a return value.

### Fail

```typescript
This rule option would have no new fail conditions, only allow passing a case that currently produces an error.
```

### Pass

```typescript
public setValue(value) { // This currently fails because of the missing return type annotation.
  this.value = value;
}
```

**Comment by HaoboZ** [comment]
The point of this rule is to require explicit types, which wants you to write return types for every function to make it clear that you didn't forget and wanted to not have a return value instead of being inferred. 
```
public setValue(value): void {
  this.value = value;
}
```
will work and not throw an error

**Comment by KristofferHenriksson** [comment]
The suggestion is to allow people to tweak the rule to be less restrictive and not force explicit void return types when they don't see the value in it. The rule is still valuable with forcing explicit return types for functions that return a value but allowing omitting them for functions that do not.

**Comment by HaoboZ** [comment]
I find it to be a bit counter intuitive to the description stating
> Explicit types for function return values makes it clear to any calling code what type is returned. This ensures that the return value is assigned to a variable of the correct type; or in the case where there is no return value, that the calling code doesn't try to use the undefined value when it shouldn't.

What's the difference between your suggestion and not having this rule?

**Comment by KristofferHenriksson** [comment]
The first part of the statement: "Explicit types for function return values makes it clear to any calling code what type is returned. This ensures that the return value is assigned to a variable of the correct type"

**Comment by bradzacher** [comment]
Duplicate of https://github.com/typescript-eslint/typescript-eslint/issues/3534, #50
